tests/integration/pilot/tunneling/tunneling_test.go

```diff
@@ -0,0 +1,416 @@
+//go:build integ
+// +build integ
+
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tunneling
```

Comment:
This test is too slow. See comparison to other _suites_ (note: most other suites are testing a much broader set of functionality):

```
ok      istio.io/istio/tests/integration/pilot/analysis 35.666s
ok      istio.io/istio/tests/integration/pilot/mcs/autoexport   41.160s
ok      istio.io/istio/tests/integration/pilot/revisioncmd      66.304s
ok      istio.io/istio/tests/integration/pilot/revisions        86.165s
ok      istio.io/istio/tests/integration/pilot/endpointslice    112.238s
ok      istio.io/istio/tests/integration/pilot/tunneling        231.002s
```

I didn't look into the details of why its slow, but I think some of my suggestions should help. Regardless, this seems like it should take at most ~30s to test, 4mins suggests we are doing things inefficiently

Reply:
There is no coverage added by tearing down a full Istio install and then reinstalling it with the same settings (45s for the reinstall, not counting the teardown time - https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/37968/integ-pilot-multicluster_istio/1527296645291577344/artifacts/trace.yaml). Similarly, there is no coverage gained by deploying + tearing down the same `sleep`, forward proxy, etc apps for each test (4x). We have invested 100s of hours into optimizing the test environment for performance, consistency, _and_ coverage. For example, the standard framework will automatically make sure calls are made from _all_ clusters - the current test only runs from a single cluster. https://github.com/istio/istio/wiki/Writing-Good-Integration-Tests may be helpful.

Reply:
> There is no coverage added by tearing down a full Istio install and then reinstalling it with the same settings

As I already said in another comment, I'm going to remove this suite.

> no coverage gained by deploying + tearing down the same sleep, forward proxy, etc apps for each test (4x)

In fact, it's the same as redeploying one app, because it's performed in parallel.

> We have invested 100s of hours into optimizing the test environment

Ok, fine. I didn't know that it's so critical. I splitted these tests into 4 tests for readability, but I can optimize execution time running all scenarios as subtests in a single test. To do that, Envoy must be run with 4 listeners (one for each case), so I have to refactor generating configuration. But I can't promise that it will not exceed 30 seconds after changes.

> https://github.com/istio/istio/wiki/Writing-Good-Integration-Tests may be helpful.

Thanks, I didn't know this wiki.

Reply:
>  I splitted these tests into 4 tests for readability,

You can have multiple tests sharing the same underlying resources. For example, in pilot/ folder ~all tests are using the same Istio deployment and echo deployments, but there are a bunch of different tests. For many of these the tests just apply some config and make few calls so take just a couple seconds.

> But I can't promise that it will not exceed 30 seconds after changes.

I don't mind this test taking a bit of time since it is testing a fairly complex topology that does has necessary setup which will add time. My concern is just keeping the _unnecessary_ setup to a minimum

Reply:
I removed suite for tunneling tests and reused echo deployments. Now the tests take ~38-40 seconds on my computer:
```
--- PASS: TestTunnelingOutboundTraffic (39.33s)
```